File: app/src/main/java/com/pr0gramm/app/ui/views/viewer/video/SurfaceViewBackend.java
package com.pr0gramm.app.ui.views.viewer.video;

import android.content.Context;
import android.view.Surface;
import android.view.SurfaceHolder;
import android.view.SurfaceView;
import android.view.View;

/**
 */
public class SurfaceViewBackend implements ViewBackend, SurfaceHolder.Callback {
    private final Callbacks callbacks;
    private final SurfaceView surfaceView;

    public SurfaceViewBackend(Context context, Callbacks callbacks) {
        this.callbacks = callbacks;

        this.surfaceView = new SurfaceView(context);
        this.surfaceView.getHolder().addCallback(this);
    }

    @Override
    public void surfaceCreated(SurfaceHolder holder) {
        callbacks.onAvailable(this);
    }

    @Override
    public void surfaceChanged(SurfaceHolder holder, int format, int width, int height) {
        callbacks.onSizeChanged(this, width, height);
    }

    @Override
    public void surfaceDestroyed(SurfaceHolder holder) {
        callbacks.onDestroy(this);
    }

    @Override
    public void setSize(int width, int height) {
        surfaceView.getHolder().setFixedSize(width, height);
    }

    @Override
    public Surface getCurrentSurface() {
        return surfaceView.getHolder().getSurface();
    }

    @Override
    public View getView() {
        return surfaceView;
    }

    @Override
    public boolean hasSurface() {
        Surface surface = surfaceView.getHolder().getSurface();
        return surface != null && surface.isValid();
    }
}

File: app/src/main/java/com/pr0gramm/app/ui/views/viewer/video/SurfaceViewBackend.kt
package com.pr0gramm.app.ui.views.viewer.video

import android.content.Context
import android.view.Surface
import android.view.SurfaceHolder
import android.view.SurfaceView
import android.view.View

/**
 */
internal class SurfaceViewBackend(
        context: Context, private val callbacks: ViewBackend.Callbacks) : ViewBackend, SurfaceHolder.Callback {

    private val surfaceView: SurfaceView = SurfaceView(context)

    init {
        this.surfaceView.holder.addCallback(this)
    }

    override val currentSurface: Surface?
        get() = surfaceView.holder.surface

    override val hasSurface: Boolean
        get() = surfaceView.holder.surface?.isValid ?: false

    override var size = ViewBackend.Size(0, 0)
        set(value) {
            field = value
            surfaceView.holder.setFixedSize(value.width, value.height)
        }

    override val view: View get() = surfaceView

    override fun surfaceCreated(holder: SurfaceHolder) {
        callbacks.onAvailable(this)
    }

    override fun surfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int) {
        callbacks.onSizeChanged(this, width, height)
    }

    override fun surfaceDestroyed(holder: SurfaceHolder) {
        callbacks.onDestroy(this)
    }

}
